File: Sources/NotchleCore/Artwork/ArtworkResolver.swift
import Foundation

/// Finds a track's album cover through Spotify's public oEmbed endpoint (no auth):
/// `GET https://open.spotify.com/oembed?url=https://open.spotify.com/track/<id>` returns JSON
/// whose `thumbnail_url` is a 300×300 cover.
///
/// Spoiler rule: callers resolve a track only once its answer is on screen (correct/revealed,
/// i.e. when `.recordOutcome` fires); never prefetch a track that is still being guessed.
public actor ArtworkResolver {
    private let http: HTTPClient
    /// Successful lookups only; a failure is retried next time.
    private var cache: [String: URL] = [:]

    public init(http: HTTPClient = URLSessionHTTPClient()) {
        self.http = http
    }

    /// The cover URL, or nil on any failure (network, status, bad JSON). Never throws.
    public func artworkURL(forTrackID trackID: String) async -> URL? {
        if let hit = cache[trackID] { return hit }
        guard let request = Self.oEmbedURL(trackID: trackID),
              let (status, body) = try? await http.get(request, headers: ["Accept": "application/json"]),
              (200..<300).contains(status),
              let url = Self.parseOEmbed(body)
        else { return nil }
        cache[trackID] = url
        return url
    }

    /// The oEmbed request for a track id (base62; anything else is rejected).
    public static func oEmbedURL(trackID: String) -> URL? {
        guard !trackID.isEmpty, trackID.allSatisfy({ $0.isASCII && ($0.isLetter || $0.isNumber) }) else { return nil }
        var c = URLComponents(string: "https://open.spotify.com/oembed")
        c?.queryItems = [URLQueryItem(name: "url", value: "https://open.spotify.com/track/\(trackID)")]
        return c?.url
    }

    /// `thumbnail_url` from an oEmbed JSON body, if it is an http(s) URL.
    public static func parseOEmbed(_ data: Data) -> URL? {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let string = object["thumbnail_url"] as? String,
              let url = URL(string: string),
              let scheme = url.scheme?.lowercased(), scheme == "https" || scheme == "http"
        else { return nil }
        return url
    }
}

File: windows/src/Notchle.Core/Artwork/ArtworkResolver.cs
using System.Collections.Concurrent;
using System.Text.Json;

namespace Notchle.Core;

/// Finds a track's album cover (Tren, 2026-09-24: "i'd like the album cover included when you
/// see the actual song"). Mirror of Sources/NotchleCore/Artwork/ArtworkResolver.swift.
/// Source: Spotify's public oEmbed endpoint (no auth),
/// <c>GET https://open.spotify.com/oembed?url=https://open.spotify.com/track/&lt;id&gt;</c>, whose
/// <c>thumbnail_url</c> is the 300×300 cover.
///
/// SPOILER RULE: callers resolve only once the track's outcome is decided (Correct / Revealed),
/// never while it is being guessed: the cover gives the album away.
public interface IArtworkResolver
{
    /// The cover's URL, or null when there is none or it could not be fetched. Never throws
    /// (except for cancellation).
    Task<Uri?> ResolveAsync(string trackId, CancellationToken cancellationToken = default);
}

public sealed class ArtworkResolver(HttpClient http) : IArtworkResolver
{
    /// Successful lookups only: a failure (offline) is retried next time.
    private readonly ConcurrentDictionary<string, Uri> _cache = new(StringComparer.Ordinal);

    public static Uri OEmbedUrl(string trackId) =>
        new("https://open.spotify.com/oembed?url=" + Uri.EscapeDataString($"https://open.spotify.com/track/{trackId}"));

    /// <c>thumbnail_url</c> of an oEmbed response, when it is an absolute http(s) URL.
    public static Uri? ParseOEmbed(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            return doc.RootElement.ValueKind == JsonValueKind.Object
                   && doc.RootElement.TryGetProperty("thumbnail_url", out var url)
                   && url.ValueKind == JsonValueKind.String
                   && Uri.TryCreate(url.GetString(), UriKind.Absolute, out var uri)
                   && uri.Scheme is "https" or "http"
                ? uri
                : null;
        }
        catch (JsonException)
        {
            return null;
        }
    }

    public async Task<Uri?> ResolveAsync(string trackId, CancellationToken cancellationToken = default)
    {
        if (string.IsNullOrWhiteSpace(trackId)) return null;
        if (_cache.TryGetValue(trackId, out var cached)) return cached;
        try
        {
            using var response = await http.GetAsync(OEmbedUrl(trackId), cancellationToken).ConfigureAwait(false);
            if (!response.IsSuccessStatusCode) return null;
            var json = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
            if (ParseOEmbed(json) is not { } uri) return null;
            _cache[trackId] = uri;
            return uri;
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception error) when (error is HttpRequestException or OperationCanceledException or IOException)
        {
            return null;
        }
    }
}
